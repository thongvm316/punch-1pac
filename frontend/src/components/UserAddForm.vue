<template>
  <form class="setting-form">
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.name.$error || errors.name }"
    >
      <label class="form-label">{{ $t('label.name') }}</label>
      <input
        v-model="$v.params.name.$model"
        class="form-input"
        type="text"
        :placeholder="$t('placeholder.name')"
      >
      <p
        v-if="$v.params.name.$error && !errors.name"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.name') }) }}
      </p>
      <p
        v-if="errors.name"
        class="form-input-hint"
      >
        {{ $t('label.name') }} {{ errors.name[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.email.$anyError || errors.email }"
    >
      <label class="form-label">{{ $t('label.email') }}</label>
      <input
        v-model="$v.params.email.$model"
        class="form-input"
        type="text"
        :placeholder="$t('placeholder.email')"
      >
      <p
        v-if="$v.params.email.$anyError && !errors.email"
        class="form-input-hint"
      >
        <span v-if="!$v.params.email.required">{{ $t('validation.required', { name: $t('label.email') }) }}</span>
        <span v-if="!$v.params.email.email">{{ $t('validation.invalid', { name: $t('label.email') }) }}</span>
      </p>
      <p
        v-if="errors.email"
        class="form-input-hint"
      >
        {{ $t('label.email') }} {{ errors.email[0] }}
      </p>
    </div>
    <div class="form-group">
      <label class="form-label">{{ $t('label.role') }}</label>
      <select
        v-model="params.role"
        class="form-select"
      >
        <option
          v-for="(role, key) in meta.roles"
          :key="key"
          :value="role"
        >
          {{ $t(`meta.roles.${role}`) }}
        </option>
      </select>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.group_id.$error || errors.group }"
    >
      <label class="form-label">{{ $t('label.group') }}</label>
      <group-select v-model="$v.params.group_id.$model">
        <option
          slot="placeholder"
          value=""
        >
          {{ $t('placeholder.chooseGroup') }}
        </option>
      </group-select>
      <p
        v-if="$v.params.group_id.$error && !errors.group"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.group') }) }}
      </p>
      <p
        v-if="errors.group"
        class="form-input-hint"
      >
        {{ $t('label.group') }} {{ errors.group[0] }}
      </p>
    </div>
    <div class="form-group">
      <button
        type="button"
        class="btn btn-success btn-submit"
        :disabled="isDisabled"
        @click="create()"
      >
        {{ $t('button.common.submit') }}
      </button>
    </div>
  </form>
</template>

<script>
import { mapState, mapActions } from 'vuex'
import handleSuccess from '../mixins/handle-success'
import userAddValidate from '../validations/user-add-validate'
const GroupSelect = () => import('./GroupSelect.vue')

export default {
  name: 'UserAddForm',

  mixins: [handleSuccess, userAddValidate],

  data() {
    return {
      errors: {},
      params: {
        name: '',
        email: '',
        role: 'member',
        group_id: ''
      }
    }
  },

  computed: {
    ...mapState('initialStates', ['meta'])
  },

  components: {
    GroupSelect
  },

  methods: {
    ...mapActions('companyUsers', ['createUser']),

    create() {
      this.createUser(this.params)
        .then(response => {
          this.handleSuccess({ message: this.$t('messages.user.addSuccess'), emitType: 'afterAdded' })
        })
        .catch(error => {
          if (error.response && error.response.status === 422) this.errors = error.response.data.errors
          else throw error
        })
    }
  }
}
</script>
