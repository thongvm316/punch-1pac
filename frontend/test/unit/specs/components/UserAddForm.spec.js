import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import setComputed from '../../supports/set-computed'
import usersData from '../../supports/fixtures/users.api'
import initialStatesData from '../../supports/fixtures/initial-states.api'
import UserAddForm from '@/components/UserAddForm'

const create = jest.spyOn(UserAddForm.methods, 'create')
const createUser = jest.spyOn(UserAddForm.methods, 'createUser').mockResolvedValue(null)
const handleSuccess = jest.spyOn(UserAddForm.mixins[0].methods, 'handleSuccess')
const CLEAR_USER_ERRORS = jest.spyOn(UserAddForm.methods, 'CLEAR_USER_ERRORS')

const localWrapperOps = {
  ...wrapperOps,
  stubs: {
    GroupSelect: true
  },
  computed: {
    isDisabled() {
      return false
    },
    meta() {
      return initialStatesData.meta
    }
  }
}

describe('UserAddForm.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(UserAddForm, localWrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when UserAddForm was mouted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(CLEAR_USER_ERRORS).toHaveBeenCalled()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when errors', () => {
    it('should match snapshot', async () => {
      setComputed(wrapper, {
        errors: usersData.errors
      })
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when create toHaveBeenCalled', () => {
    it('should create method called', async () => {
      wrapper.find('.btn.btn-success.btn-submit').trigger('click')
      await wrapper.vm.$nextTick()

      expect(create).toHaveBeenCalled()
      expect(createUser).toHaveBeenCalledWith(wrapper.vm.params)
      expect(handleSuccess).toHaveBeenCalledWith({
        message: 'An user is created. Email contains login information is sent to user',
        emitType: 'afterAdded'
      })
    })
  })
})
