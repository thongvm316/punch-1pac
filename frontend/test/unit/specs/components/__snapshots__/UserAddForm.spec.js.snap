// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAddForm.vue when UserAddForm was mouted should render correctly 1`] = `
<form class="setting-form">
  <div class="form-group"><label class="form-label">Name</label> <input type="text" placeholder="Name" class="form-input">
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Email</label> <input type="text" placeholder="Email" class="form-input">
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Role</label> <select class="form-select">
      <option value="member">
        Member
      </option>
      <option value="admin">
        Admin
      </option>
      <option value="superadmin">
        Superadmin
      </option>
    </select></div>
  <div class="form-group"><label class="form-label">Group</label>
    <groupselect-stub value="">
      <option value="">
        Choose a group
      </option>
    </groupselect-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Submit
    </button></div>
</form>
`;

exports[`UserAddForm.vue when errors should match snapshot 1`] = `
<form class="setting-form">
  <div class="form-group has-error"><label class="form-label">Name</label> <input type="text" placeholder="Name" class="form-input">
    <!---->
    <p class="form-input-hint">
      Name to have been taken
    </p>
  </div>
  <div class="form-group has-error"><label class="form-label">Email</label> <input type="text" placeholder="Email" class="form-input">
    <!---->
    <p class="form-input-hint">
      Email to have been taken
    </p>
  </div>
  <div class="form-group"><label class="form-label">Role</label> <select class="form-select">
      <option value="member">
        Member
      </option>
      <option value="admin">
        Admin
      </option>
      <option value="superadmin">
        Superadmin
      </option>
    </select></div>
  <div class="form-group"><label class="form-label">Group</label>
    <groupselect-stub value="">
      <option value="">
        Choose a group
      </option>
    </groupselect-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Submit
    </button></div>
</form>
`;
